### PR TITLE
Memory and object management

### PR DESCRIPTION
**Pull request by CodingRays**
Vulkan object and memory management.

Objects for a single vulkan device are managed by a `ObjectManager` instance.
Management is organized into 2 collections.
1. The lifetime of objects is determined by `ObjectSet`s. All objects in a set are created and destroyed at the same time.
2. Access to objects is controlled by `SynchronizationGroup`s. Every group contains one timeline semaphore which ensures safe access to objects protected by the group. All objects in a set belong to the same synchronization group.

### CODE REVIEW

File: src/rosella.rs [hunk 1]
```diff
@@ -52,10 +54,13 @@ impl Rosella {
         let elapsed = now.elapsed();
         println!("Instance & Device Initialization took: {:.2?}", elapsed);
 
+        let object_manager = ObjectManager::new(device.clone());
```
Comment (hYdos) [inline src/rosella.rs:55]: Reminds me of Java Rosella

**Review by hYdos — APPROVED** [review]

**Comment by CodingRays** [comment]
I merged my current work on the improved object management.
This only includes clean-up refactorings and some critical bugfixes (this is the reason i updated this PR instead of waiting).

File: src/util/slice_splitter.rs [hunk 2]
```diff
@@ -0,0 +1,28 @@
+pub struct Splitter<'a, T> {
```
Comment (ramidzkh) [inline src/util/slice_splitter.rs:3]: Since it's pub, could this be documented?

**Review by ramidzkh — APPROVED** [review]

**Review by hYdos — APPROVED** [review]
I still approve but getting that thing that ramid pointed out document would be good

**Pull request merged**